### PR TITLE
CorsHandler for Cross-origin resource sharing

### LINKED ISSUE #453: Add CorsHandler
What do you think of adding a handler for CORS? Analogous to the existing CSRFHandler.

I had the need and I made two implementations.

How can I submit them to decide which one to use? (if there is interest)

### PR DESCRIPTION
**Pull request by mhagnumdw**
Resolve #453 

### D R A F T

Handler to define how CORS requests are handled.

Reason for DRAFT: I made two implementations: CorsHandler and CorsHandlerV2 (simpler code).

`I strongly recommend CorsHandler.`

### CorsHandler

```java
final CorsHandler corsHandler =
    CorsHandler.builder()
        .addAllowOrigin("http://localhost:4200")
        .addAllowMethod(HttpConstants.Method.GET)
        .addAllowMethod(HttpConstants.Method.POST)
        .addAllowHeader(HttpConstants.Header.AUTHORIZATION)
        .build();
ANY("/webapp/api/.*", corsHandler);
```
### CorsHandlerV2

```java
final CorsHandlerV2 corsHandler = new CorsHandlerV2("http://localhost:4200");
corsHandler.setAllowMethods("GET, POST");
corsHandler.setAllowHeaders("Authorization");
ANY("/webapp/api/.*", corsHandler);

// or

final CorsHandlerV2 corsHandler = new CorsHandlerV2("http://localhost:4200");
corsHandler.setAllowMethods(String.join(", ", HttpConstants.Method.GET, HttpConstants.Method.POST));
corsHandler.setAllowHeaders(HttpConstants.Header.AUTHORIZATION);
ANY("/webapp/api/.*", corsHandler);
```

### CODE REVIEW

**Comment by decebals** [comment]
I think that I prefer a mix between both: variant two (no builder) is more simple/short, with fluent methods (return this), without `set` prefix (for example `allowMethod(s)` instead of `setAllowMethod(s)`) and with a out of the box configuration where is possible (probably only url in constructor is required).

**Comment by mhagnumdw** [comment]
@decebals, following your proposal, I added a new implementation: CorsHandlerV3

How to use
```java
final CorsHandlerV3 corsHandler = 
    new CorsHandlerV3("http://localhost:4200")
        .allowMethods("GET, POST")
        .allowHeaders("Authorization");
ANY("/webapp/api/.*", corsHandler);

// or

final CorsHandlerV3 corsHandler =
    new CorsHandlerV3("http://localhost:4200")
        .setAllowMethods(String.join(", ", HttpConstants.Method.GET, HttpConstants.Method.POST))
        .setAllowHeaders(HttpConstants.Header.AUTHORIZATION);
ANY("/webapp/api/.*", corsHandler);
```

Anything let me know.
If this is ok, I will delete the other two and rename it from CorsHandlerV3 to CorsHandler.

File: pippo-core/src/main/java/ro/pippo/core/route/CorsHandlerV3.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    public CorsHandlerV3(String allowOrigin) {
+        if (StringUtils.isNullOrEmpty(allowOrigin)) {
+            throw new PippoRuntimeException("allowOrigin cannot be blank");
+        }
+        this.allowOrigin = allowOrigin;
+    }
+
+    @Override
+    public void handle(RouteContext context) {
+        context.getResponse().header(HttpConstants.Header.ACCESS_CONTROL_ALLOW_ORIGIN, allowOrigin);
```
Comment (decebals) [inline pippo-core/src/main/java/ro/pippo/core/route/CorsHandlerV3.java]: Maybe here to extract the response object and to use it in next lines:
```java
Response response = context.getResponse();
```

**Comment by decebals** [comment]
@mhagnumdw 
Looks good. Only am minor comment (it's not something serious).
For me it's important that you also be happy with what went out.

File: pippo-core/src/main/java/ro/pippo/core/route/CorsHandlerV3.java (hunk 1 above)
Comment (mhagnumdw) [reply pippo-core/src/main/java/ro/pippo/core/route/CorsHandlerV3.java]: Done!

**Comment by mhagnumdw** [comment]
> @mhagnumdw
> Looks good. Only am minor comment (it's not something serious).
> For me it's important that you also be happy with what went out.

I'm happy!
I'm currently using this handler (Builder version) in my application. I'll change it with this new handler's version as soon as possible.

**Comment by decebals** [comment]
That is all. Rename `CorsHandlerV3` to `CorsHandler` and we can merge (with squash) the new file.

**Comment by mhagnumdw** [comment]
Done.

**Comment by decebals** [comment]
Thanks!

**Pull request merged**